File: scripts/transcribe.py
#!/usr/bin/env python3
"""
Transcribe a local media file or a YouTube URL into a plain-text transcript,
using the OpenAI Whisper API directly (no Java/server dependency).

Usage:
    python scripts/transcribe.py --project <name> --file <path> [--language en]
    python scripts/transcribe.py --project <name> --youtube-url <url> [--language en]
    python scripts/transcribe.py --project <name> --file <path> --out <custom/path.md>

Writes the transcript directly to disk (default:
projects/<project>/05_transcription_transcript.md) and prints a short summary to
stdout. Never prints the transcript itself to stdout.
"""
import argparse
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

import requests

from common import openai_config, project_dir, fail

OPENAI_MAX_AUDIO_BYTES = 25 * 1024 * 1024
OPENAI_CHUNK_SECONDS = 540
YOUTUBE_EXTRACTOR_ARGS = "youtube:player_client=android,web"
YTDLP_TIMEOUT_SECONDS = 15 * 60
FFMPEG_TIMEOUT_SECONDS = 10 * 60
WHISPER_TIMEOUT_SECONDS = 5 * 60


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--project", required=True, help="Project name under projects/")
    source = parser.add_mutually_exclusive_group(required=True)
    source.add_argument("--file", help="Path to a local media file (audio or video)")
    source.add_argument("--youtube-url", help="A YouTube URL to download and transcribe")
    parser.add_argument("--language", default=None, help="ISO language code hint, e.g. en")
    parser.add_argument("--out", default=None, help="Output path (default: projects/<project>/05_transcription_transcript.md)")
    return parser.parse_args()


def run(command, timeout, log_path, description):
    with open(log_path, "wb") as log_file:
        try:
            process = subprocess.run(
                command,
                stdout=log_file,
                stderr=subprocess.STDOUT,
                timeout=timeout,
            )
        except subprocess.TimeoutExpired:
            fail(f"{description} timed out after {timeout}s")
        except FileNotFoundError:
            fail(f"{description} failed: '{command[0]}' not found on PATH")
    if process.returncode != 0:
        tail = log_tail(log_path)
        fail(f"{description} failed (exit {process.returncode}):\n{tail}")


def log_tail(log_path, lines=40):
    try:
        text = Path(log_path).read_text(encoding="utf-8", errors="replace")
    except OSError as exc:
        return f"<failed to read log: {exc}>"
    return "\n".join(text.splitlines()[-lines:])


def download_youtube_audio(url: str, tmp_dir: Path) -> Path:
    if shutil.which("yt-dlp") is None:
        fail("yt-dlp is not on PATH. Install it (pip install yt-dlp) and retry.")
    log_path = tmp_dir / "yt-dlp.log"
    command = [
        "yt-dlp",
        "--no-update",
        "--no-progress",
        "--retries", "3",
        "--fragment-retries", "3",
        "--retry-sleep", "2",
        "--force-ipv4",
        "--js-runtimes", "node,deno",
        "--extractor-args", YOUTUBE_EXTRACTOR_ARGS,
        "-x",
        "--audio-format", "mp3",
        "-o", str(tmp_dir / "%(id)s.%(ext)s"),
        url,
    ]
    run(command, YTDLP_TIMEOUT_SECONDS, log_path, "yt-dlp download")
    downloaded = [p for p in tmp_dir.iterdir() if p.name not in ("yt-dlp.log",) and p.is_file()]
    if not downloaded:
        fail("yt-dlp reported success but produced no audio file")
    downloaded.sort(key=lambda p: p.stat().st_mtime, reverse=True)
    return downloaded[0]


def chunk_audio(input_path: Path, tmp_dir: Path) -> list[Path]:
    if shutil.which("ffmpeg") is None:
        fail("ffmpeg is not on PATH. Install it and retry (needed to split audio over 25MB).")
    chunk_dir = tmp_dir / "chunks"
    chunk_dir.mkdir(exist_ok=True)
    log_path = tmp_dir / "ffmpeg-split.log"
    output_pattern = chunk_dir / "chunk-%03d.mp3"
    command = [
        "ffmpeg",
        "-hide_banner",
        "-loglevel", "error",
        "-y",
        "-i", str(input_path),
        "-vn",
        "-ac", "1",
        "-ar", "16000",
        "-codec:a", "libmp3lame",
        "-b:a", "48k",
        "-f", "segment",
        "-segment_time", str(OPENAI_CHUNK_SECONDS),
        str(output_pattern),
    ]
    run(command, FFMPEG_TIMEOUT_SECONDS, log_path, "ffmpeg chunking")
    chunks = sorted(chunk_dir.glob("chunk-*.mp3"))
    if not chunks:
        fail("ffmpeg reported success but produced no chunk files")
    return chunks


def transcribe_chunk(cfg: dict, data: bytes, filename: str, language: str | None) -> list[dict]:
    if len(data) > OPENAI_MAX_AUDIO_BYTES:
        fail(f"{filename} is {len(data)} bytes, still over the 25MB OpenAI limit after chunking")
    fields = {
        "model": (None, cfg["transcription_model"]),
        "response_format": (None, "verbose_json"),
        "timestamp_granularities[]": (None, "segment"),
        "file": (filename, data, "audio/mpeg"),
    }
    if language:
        fields["language"] = (None, language)
    url = f"{cfg['base_url']}/v1/audio/transcriptions"
    try:
        response = requests.post(
            url,
            headers={"Authorization": f"Bearer {cfg['api_key']}"},
            files=fields,
            timeout=WHISPER_TIMEOUT_SECONDS,
        )
    except requests.RequestException as exc:
        fail(f"OpenAI transcription request failed: {exc}")
    if response.status_code >= 400:
        fail(f"OpenAI transcription failed (HTTP {response.status_code}): {extract_error(response)}")
    body = response.json()
    segments = body.get("segments")
    if not segments:
        text = (body.get("text") or "").strip()
        return [{"start": 0.0, "end": 0.0, "speaker": "Unknown", "text": text}] if text else []
    entries = []
    for segment in segments:
        text = (segment.get("text") or "").strip()
        if not text:
            continue
        entries.append({
            "start": float(segment.get("start", 0.0)),
            "end": float(segment.get("end", segment.get("start", 0.0))),
            "speaker": segment.get("speaker", "Unknown"),
            "text": text,
        })
    return entries


def extract_error(response) -> str:
    try:
        body = response.json()
        return body.get("error", {}).get("message", response.text[:300])
    except ValueError:
        return response.text[:300]


def format_timestamp(seconds_value: float) -> str:
    total_seconds = max(0, int(seconds_value // 1))
    hours, remainder = divmod(total_seconds, 3600)
    minutes, seconds = divmod(remainder, 60)
    if hours > 0:
        return f"{hours:02d}:{minutes:02d}:{seconds:02d}"
    return f"{minutes:02d}:{seconds:02d}"


def render_transcript(entries: list[dict]) -> str:
    lines = [
        f"[{format_timestamp(e['start'])} - {format_timestamp(e['end'])}] {e['speaker']}: {e['text']}"
        for e in entries
    ]
    return "\n".join(lines) + ("\n" if lines else "")


def main():
    args = parse_args()
    cfg = openai_config()
    proj_dir = project_dir(args.project)
    out_path = Path(args.out) if args.out else proj_dir / "05_transcription_transcript.md"

    with tempfile.TemporaryDirectory(prefix="stream-producer-transcribe-") as tmp:
        tmp_dir = Path(tmp)

        if args.youtube_url:
            audio_path = download_youtube_audio(args.youtube_url, tmp_dir)
        else:
            audio_path = Path(args.file)
            if not audio_path.is_file():
                fail(f"File not found: {audio_path}")

        size = audio_path.stat().st_size
        if size > OPENAI_MAX_AUDIO_BYTES:
            chunk_paths = chunk_audio(audio_path, tmp_dir)
        else:
            chunk_paths = [audio_path]

        all_entries = []
        for index, chunk_path in enumerate(chunk_paths):
            data = chunk_path.read_bytes()
            entries = transcribe_chunk(cfg, data, chunk_path.name, args.language)
            offset = index * OPENAI_CHUNK_SECONDS
            for entry in entries:
                entry["start"] += offset
                entry["end"] += offset
            all_entries.extend(entries)

    if not all_entries:
        fail("Transcription produced no text. Check the input file/URL and try again.")

    transcript_text = render_transcript(all_entries)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(transcript_text, encoding="utf-8")

    print(f"Transcribed {len(all_entries)} segments ({len(transcript_text)} chars) -> {out_path}")


if __name__ == "__main__":
    main()

File: scripts/transcribe.js
#!/usr/bin/env node
// Transcribe a local media file or a YouTube URL into a plain-text transcript,
// using the OpenAI Whisper API directly (no Python dependency).
//
// Usage:
//   node scripts/transcribe.js --project <name> --file <path> [--language en]
//   node scripts/transcribe.js --project <name> --youtube-url <url> [--language en]
//   node scripts/transcribe.js --project <name> --file <path> --out <custom/path.md>
//
// Writes the transcript directly to disk (default:
// projects/<project>/05_transcription_transcript.md) and prints a short summary to
// stdout. Never prints the transcript itself to stdout.
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { parseArgs } from 'node:util';
import { openaiConfig, projectDir, fail } from './common.js';

const OPENAI_MAX_AUDIO_BYTES = 25 * 1024 * 1024;
const OPENAI_CHUNK_SECONDS = 540;
const YOUTUBE_EXTRACTOR_ARGS = 'youtube:player_client=android,web';
const YTDLP_TIMEOUT_MS = 15 * 60 * 1000;
const FFMPEG_TIMEOUT_MS = 10 * 60 * 1000;
const WHISPER_TIMEOUT_MS = 5 * 60 * 1000;

function parseCliArgs() {
  const { values } = parseArgs({
    options: {
      project: { type: 'string' },
      file: { type: 'string' },
      'youtube-url': { type: 'string' },
      language: { type: 'string' },
      out: { type: 'string' },
    },
  });

  if (!values.project) {
    fail('--project is required.');
  }
  // XOR: exactly one of --file / --youtube-url must be given.
  if (!values.file === !values['youtube-url']) {
    fail('Exactly one of --file or --youtube-url is required.');
  }

  return {
    project: values.project,
    file: values.file ?? null,
    youtubeUrl: values['youtube-url'] ?? null,
    language: values.language ?? null,
    out: values.out ?? null,
  };
}

function commandExists(cmd) {
  const checker = process.platform === 'win32' ? 'where' : 'which';
  const result = spawnSync(checker, [cmd], { encoding: 'utf8' });
  return result.status === 0;
}

function tail(text, lines) {
  return text.split('\n').slice(-lines).join('\n');
}

/** Runs a subprocess synchronously, capturing combined stdout+stderr in memory
 *  (no temp log file needed — the tempdir this runs inside is discarded either way). */
function run(command, args, timeoutMs, description) {
  const result = spawnSync(command, args, { timeout: timeoutMs, encoding: 'utf8' });

  if (result.error) {
    if (result.error.code === 'ENOENT') {
      fail(`${description} failed: '${command}' not found on PATH`);
    }
    if (result.error.code === 'ETIMEDOUT') {
      fail(`${description} timed out after ${Math.round(timeoutMs / 1000)}s`);
    }
    fail(`${description} failed: ${result.error.message}`);
  }

  const combined = `${result.stdout ?? ''}${result.stderr ?? ''}`;
  if (result.status !== 0) {
    fail(`${description} failed (exit ${result.status}):\n${tail(combined, 40)}`);
  }
}

function downloadYoutubeAudio(url, tmpDir) {
  if (!commandExists('yt-dlp')) {
    fail('yt-dlp is not on PATH. Install it (pip install yt-dlp) and retry.');
  }
  const args = [
    '--no-update', '--no-progress', '--retries', '3', '--fragment-retries', '3',
    '--retry-sleep', '2', '--force-ipv4', '--js-runtimes', 'node,deno',
    '--extractor-args', YOUTUBE_EXTRACTOR_ARGS, '-x', '--audio-format', 'mp3',
    '-o', path.join(tmpDir, '%(id)s.%(ext)s'), url,
  ];
  run('yt-dlp', args, YTDLP_TIMEOUT_MS, 'yt-dlp download');

  const downloaded = fs.readdirSync(tmpDir)
    .map((name) => path.join(tmpDir, name))
    .filter((p) => fs.statSync(p).isFile());
  if (downloaded.length === 0) {
    fail('yt-dlp reported success but produced no audio file');
  }
  downloaded.sort((a, b) => fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs);
  return downloaded[0];
}

function chunkAudio(inputPath, tmpDir) {
  if (!commandExists('ffmpeg')) {
    fail('ffmpeg is not on PATH. Install it and retry (needed to split audio over 25MB).');
  }
  const chunkDir = path.join(tmpDir, 'chunks');
  fs.mkdirSync(chunkDir, { recursive: true });
  const outputPattern = path.join(chunkDir, 'chunk-%03d.mp3');
  const args = [
    '-hide_banner', '-loglevel', 'error', '-y', '-i', inputPath, '-vn',
    '-ac', '1', '-ar', '16000', '-codec:a', 'libmp3lame', '-b:a', '48k',
    '-f', 'segment', '-segment_time', String(OPENAI_CHUNK_SECONDS), outputPattern,
  ];
  run('ffmpeg', args, FFMPEG_TIMEOUT_MS, 'ffmpeg chunking');

  const chunks = fs.readdirSync(chunkDir)
    .filter((name) => name.startsWith('chunk-') && name.endsWith('.mp3'))
    .sort()
    .map((name) => path.join(chunkDir, name));
  if (chunks.length === 0) {
    fail('ffmpeg reported success but produced no chunk files');
  }
  return chunks;
}

async function extractError(response) {
  const raw = await response.text();
  try {
    const body = JSON.parse(raw);
    return body?.error?.message ?? raw.slice(0, 300);
  } catch {
    return raw.slice(0, 300);
  }
}

async function transcribeChunk(cfg, data, filename, language) {
  if (data.length > OPENAI_MAX_AUDIO_BYTES) {
    fail(`${filename} is ${data.length} bytes, still over the 25MB OpenAI limit after chunking`);
  }

  const form = new FormData();
  form.set('model', cfg.transcriptionModel);
  form.set('response_format', 'verbose_json');
  form.set('timestamp_granularities[]', 'segment');
  form.set('file', new Blob([data], { type: 'audio/mpeg' }), filename);
  if (language) {
    form.set('language', language);
  }

  const url = `${cfg.baseUrl}/v1/audio/transcriptions`;
  let response;
  try {
    response = await fetch(url, {
      method: 'POST',
      headers: { Authorization: `Bearer ${cfg.apiKey}` },
      body: form,
      signal: AbortSignal.timeout(WHISPER_TIMEOUT_MS),
    });
  } catch (err) {
    fail(`OpenAI transcription request failed: ${err.message}`);
  }

  if (response.status >= 400) {
    fail(`OpenAI transcription failed (HTTP ${response.status}): ${await extractError(response)}`);
  }

  const body = await response.json();
  const segments = body.segments;
  if (!segments || segments.length === 0) {
    const text = (body.text ?? '').trim();
    return text ? [{ start: 0.0, end: 0.0, speaker: 'Unknown', text }] : [];
  }

  const entries = [];
  for (const segment of segments) {
    const text = (segment.text ?? '').trim();
    if (!text) continue;
    const start = Number(segment.start ?? 0.0);
    entries.push({
      start,
      end: Number(segment.end ?? start),
      speaker: segment.speaker ?? 'Unknown',
      text,
    });
  }
  return entries;
}

function formatTimestamp(secondsValue) {
  const totalSeconds = Math.max(0, Math.floor(secondsValue));
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;
  const pad = (n) => String(n).padStart(2, '0');
  if (hours > 0) {
    return `${pad(hours)}:${pad(minutes)}:${pad(seconds)}`;
  }
  return `${pad(minutes)}:${pad(seconds)}`;
}

function renderTranscript(entries) {
  const lines = entries.map(
    (e) => `[${formatTimestamp(e.start)} - ${formatTimestamp(e.end)}] ${e.speaker}: ${e.text}`,
  );
  return lines.length ? lines.join('\n') + '\n' : '';
}

async function main() {
  const args = parseCliArgs();
  const cfg = openaiConfig();
  const projDir = projectDir(args.project);
  const outPath = args.out
    ? path.resolve(args.out)
    : path.join(projDir, '05_transcription_transcript.md');

  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'stream-helper-skills-transcribe-'));
  const allEntries = [];
  try {
    let audioPath;
    if (args.youtubeUrl) {
      audioPath = downloadYoutubeAudio(args.youtubeUrl, tmpDir);
    } else {
      audioPath = path.resolve(args.file);
      if (!fs.existsSync(audioPath) || !fs.statSync(audioPath).isFile()) {
        fail(`File not found: ${audioPath}`);
      }
    }

    const size = fs.statSync(audioPath).size;
    const chunkPaths = size > OPENAI_MAX_AUDIO_BYTES ? chunkAudio(audioPath, tmpDir) : [audioPath];

    for (let index = 0; index < chunkPaths.length; index += 1) {
      const chunkPath = chunkPaths[index];
      const data = fs.readFileSync(chunkPath);
      const entries = await transcribeChunk(cfg, data, path.basename(chunkPath), args.language);
      const offset = index * OPENAI_CHUNK_SECONDS;
      for (const entry of entries) {
        entry.start += offset;
        entry.end += offset;
      }
      allEntries.push(...entries);
    }
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }

  if (allEntries.length === 0) {
    fail('Transcription produced no text. Check the input file/URL and try again.');
  }

  const transcriptText = renderTranscript(allEntries);
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, transcriptText, 'utf8');
  console.log(
    `Transcribed ${allEntries.length} segments (${transcriptText.length} chars) -> ${outPath}`,
  );
}

main().catch((err) => {
  fail(err?.message ?? String(err));
});
